sql/core/src/test/scala/org/apache/spark/sql/streaming/test/DataStreamReaderWriterSuite.scala

```diff
@@ -467,4 +468,54 @@ class DataStreamReaderWriterSuite extends StreamTest with BeforeAndAfter {
     val sq = df.writeStream.format("console").start()
     sq.stop()
   }
+
+  test("MemorySink can recover from a checkpoint in Complete Mode") {
+    import testImplicits._
+    val ms = new MemoryStream[Int](0, sqlContext)
+    val df = ms.toDF().toDF("a")
+    val checkpointLoc = newMetadataDir
+    val checkpointDir = new File(checkpointLoc, "offsets")
+    checkpointDir.mkdirs()
+    assert(checkpointDir.exists())
+    def startStream: StreamingQuery = {
+      df.groupBy("a")
+        .count()
+        .writeStream
+        .format("memory")
+        .queryName("test")
+        .option("checkpointLocation", checkpointLoc)
+        .outputMode("complete")
+        .start()
+    }
+    // no exception here
+    val q = startStream
+    ms.addData(0, 1)
+    q.processAllAvailable()
+    q.stop()
+
+    checkAnswer(
+      spark.table("test"),
+      Seq(Row(0, 1), Row(1, 1))
+    )
+
+    val q2 = startStream
+    q2.processAllAvailable()
+    checkAnswer(
+      spark.table("test"),
+      Seq(Row(0, 1), Row(1, 1))
+    )
```

Comment:
its not clear if this table was regenerated by the query, or just the same table that existed before when the query was stopped earlier. so it would be good it if before restarting you remove the table, or rewrite the table to something different (and verify that), so that we are sure that the table is actually being regenerated.